crates/teloxide-core/src/net/request.rs

```diff
@@ -100,6 +100,55 @@ where
     let text = response.text().await?;
 
     serde_json::from_str::<TelegramResponse<T>>(&text)
+        .map(|mut response| {
+            use crate::types::{Update, UpdateKind};
+            use std::{any::Any, iter::zip};
+
+            // HACK: Fill-in error information into `UpdateKind::Error`.
+            //
+            //       Why? Well, we need `Update` deserialization to be reliable,
+            //       even if Telegram breaks something in their Bot API, we want
+            //       1. Deserialization to """succeed"""
+            //       2. Get the `update.id`
+            //
+            //       Both of these points are required for `get_updates(...) -> Vec<Update>`
+            //       to behave well after Telegram introduces updates that we can't parse.
+            //       (1.) makes it so only some of the updates in a butch need to be skipped
+            //       (otherwise serde'll stop on the first error). (2.) allows us to issue
+            //       the next `get_updates` call with the right offset, even if the last
+            //       update in the batch didn't deserialize well.
+            //
+            //       serde's interface doesn't allows us to implement `Deserialize` in such
+            //       a way, that we could keep the data we couldn't parse, so our
+            //       `Deserialize` impl for `UpdateKind` just returns
+            //       `UpdateKind::Error(/* some empty-ish value */)`. Here, through some
+            //       terrible hacks and downcasting, we fill-in the data we couldn't parse
+            //       so that our users can make actionable bug reports.
+            //
+            //       We specifically handle `Vec<Update>` here, because that's the return
+            //       type of the only method that returns updates.
+            if TypeId::of::<T>() == TypeId::of::<Vec<Update>>() {
```

Comment:
This block contains some involved logic, and it'd be nice to have a unit test for it.

Reply:
Done